### libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
df = pd.read_csv('/content/salaries.csv')
df

Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651       Not provided   
148651  148652       Not provided   
148652  148653       Not provided   
148653  148654          Joe Lopez   

                                              JobTitle    BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)  155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)  212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC   77916.00   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)  134401.60   
...                                                ...        ...   
148649                                       Custodian       0.00   
148650                                    Not provided        NaN   
148651                                    Not provided        NaN   
148652                                    Not provided        NaN   
148653                      Counselor, Log Cabin Ranch       0.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0              0.00  400184.25       NaN  567595.43         567595.43  2011   
1         245131.88  137811.38       NaN  538909.28         538909.28  2011   
2         106088.18   16452.60       NaN  335279.91         335279.91  2011   
3          56120.71  198306.90       NaN  332343.61         332343.61  2011   
4           9737.00  182234.59       NaN  326373.19         326373.19  2011   
...             ...        ...       ...        ...               ...   ...   
148649         0.00       0.00       0.0       0.00              0.00  2014   
148650          NaN        NaN       NaN       0.00              0.00  2014   
148651          NaN        NaN       NaN       0.00              0.00  2014   
148652          NaN        NaN       NaN       0.00              0.00  2014   
148653         0.00    -618.13       0.0    -618.13           -618.13  2014   

        Notes         Agency  Status  
0         NaN  San Francisco     NaN  
1         NaN  San Francisco     NaN  
2         NaN  San Francisco     NaN  
3         NaN  San Francisco     NaN  
4         NaN  San Francisco     NaN  
...       ...            ...     ...  
148649    NaN  San Francisco     NaN  
148650    NaN  San Francisco     NaN  
148651    NaN  San Francisco     NaN  
148652    NaN  San Francisco     NaN  
148653    NaN  San Francisco     NaN  

[148654 rows x 13 columns]

In [ ]:
df.columns

Index(['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')

In [ ]:
df = df.drop(columns=['Id', 'Notes', 'Agency', 'Status'])

In [ ]:
df

EmployeeName                                        JobTitle  \
0          NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1            GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2          ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3       CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4         PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
...                   ...                                             ...   
148649      Roy I Tillery                                       Custodian   
148650       Not provided                                    Not provided   
148651       Not provided                                    Not provided   
148652       Not provided                                    Not provided   
148653          Joe Lopez                      Counselor, Log Cabin Ranch   

          BasePay  OvertimePay   OtherPay  Benefits   TotalPay  \
0       167411.18         0.00  400184.25       NaN  567595.43   
1       155966.02    245131.88  137811.38       NaN  538909.28   
2       212739.13    106088.18   16452.60       NaN  335279.91   
3        77916.00     56120.71  198306.90       NaN  332343.61   
4       134401.60      9737.00  182234.59       NaN  326373.19   
...           ...          ...        ...       ...        ...   
148649       0.00         0.00       0.00       0.0       0.00   
148650        NaN          NaN        NaN       NaN       0.00   
148651        NaN          NaN        NaN       NaN       0.00   
148652        NaN          NaN        NaN       NaN       0.00   
148653       0.00         0.00    -618.13       0.0    -618.13   

        TotalPayBenefits  Year  
0              567595.43  2011  
1              538909.28  2011  
2              335279.91  2011  
3              332343.61  2011  
4              326373.19  2011  
...                  ...   ...  
148649              0.00  2014  
148650              0.00  2014  
148651              0.00  2014  
148652              0.00  2014  
148653           -618.13  2014  

[148654 rows x 9 columns]

In [ ]:
df.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [ ]:
df = df.dropna()

In [ ]:
df

EmployeeName                      JobTitle    BasePay  \
36159       Gary Altenberg  Lieutenant, Fire Suppression  128808.87   
36160         Gregory Suhr               Chief of Police  302578.00   
36161           Khoa Trinh   Electronic Maintenance Tech  111921.00   
36162   Joanne Hayes-White        Chief, Fire Department  296943.01   
36163    Frederick Binkley     EMT/Paramedic/Firefighter  126863.19   
...                    ...                           ...        ...   
148645    Carolyn A Wilson     Human Services Technician       0.00   
148647      Joann Anderson   Communications Dispatcher 2       0.00   
148648         Leon Walker                     Custodian       0.00   
148649       Roy I Tillery                     Custodian       0.00   
148653           Joe Lopez    Counselor, Log Cabin Ranch       0.00   

        OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  
36159     220909.48  13126.31  44430.12  362844.66         407274.78  2012  
36160          0.00  18974.11  69810.19  321552.11         391362.30  2012  
36161     146415.32  78057.41  53102.29  336393.73         389496.02  2012  
36162          0.00  17816.59  72047.88  314759.60         386807.48  2012  
36163     192424.49  17917.18  44438.25  337204.86         381643.11  2012  
...             ...       ...       ...        ...               ...   ...  
148645         0.00      0.00      0.00       0.00              0.00  2014  
148647         0.00      0.00      0.00       0.00              0.00  2014  
148648         0.00      0.00      0.00       0.00              0.00  2014  
148649         0.00      0.00      0.00       0.00              0.00  2014  
148653         0.00   -618.13      0.00    -618.13           -618.13  2014  

[111886 rows x 9 columns]

In [ ]:
df.isnull().sum()

EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
dtype: int64

1. Find The Occurence Of the Employee Names(Top5)

In [ ]:
employee_counts = df['EmployeeName'].value_counts().head(5)
print(employee_counts)

EmployeeName
Kevin Lee       13
Steven Lee      11
William Wong    11
Richard Lee     11
John Chan        9
Name: count, dtype: int64


## Documentation

This output is a result of using the .value_counts() method on the EmployeeName column of a Pandas DataFrame. It shows the frequency of the top 5 most common employee names in the dataset.

Kevin Lee 13: This indicates that the employee name "Kevin Lee" appears 13 times in the dataset.
Steven Lee 11: This indicates that the employee name "Steven Lee" appears 11 times in the dataset.
William Wong 11: This indicates that the employee name "William Wong" appears 11 times in the dataset.
Richard Lee 11: This indicates that the employee name "Richard Lee" appears 11 times in the dataset.
John Chan 9: This indicates that the employee name "John Chan" appears 9 times in the dataset.
Name: count, dtype: int64: This part provides information about the output itself. * Name: count: This specifies that the column containing the employee names is labeled "count." * dtype: int64: This indicates that the data type of the values (the counts) is an integer of 64-bit size.

2. Find The Number Of Unique Jobs Titles


In [ ]:
num_unique_jobs = df['JobTitle'].nunique()
print(f"Number of unique job titles: {num_unique_jobs}")

Number of unique job titles: 1109


Documentation

This output means there are 1109 different job titles in the dataset. In other words, there are 1109 unique entries within the 'JobTitle' column of the DataFrame. This information provides insight into the diversity of job roles represented in the dataset.

3. Find The Total Number Of Job Titles Contain captain


In [ ]:
captain_count = df['JobTitle'].str.contains('captain', case=False).sum()
print(f"Number of job titles containing 'captain': {captain_count}")

Number of job titles containing 'captain': 410


Documentation

This output means that there are 410 job titles in the dataset that contain the word "captain" (in any case). This provides information about the prevalence of job roles related to "captain" within the dataset.

4. Display All the Employee Names from Fire Department

In [ ]:
fire_department_employees = df[df['JobTitle'].str.contains('fire', case=False)]['EmployeeName']
print(fire_department_employees)

36159         Gary Altenberg
36162     Joanne Hayes-White
36163      Frederick Binkley
36168         David Franklin
36169           Brendan Ward
                 ...        
145956      Kenneth C Farris
147556         Edward A Dunn
148021        Kari A Johnson
148209          Sheryl K Lee
148554       Lawrence F Gatt
Name: EmployeeName, Length: 4399, dtype: object


documentation

The output shows a list of employee names, along with their original index values, who belong to the Fire Department

5. Find The Minimum,Maximum and Average BasePay


In [ ]:
minimum_basepay = df['BasePay'].min()
maximum_basepay = df['BasePay'].max()
average_basepay = df['BasePay'].mean()

print(f"Minimum BasePay: {minimum_basepay}")
print(f"Maximum BasePay: {maximum_basepay}")
print(f"Average BasePay: {average_basepay}")

Minimum BasePay: -166.01
Maximum BasePay: 319275.01
Average BasePay: 67207.55842464653


Documentation

Minimum BasePay: -166.01: This indicates the lowest value present in the 'BasePay' column of the DataFrame. In this case, the minimum base pay is -166.01. A negative base pay could possibly represent deductions or specific situations within the dataset's context.
Maximum BasePay: 319275.01: This represents the highest value in the 'BasePay' column. Here, the maximum base pay is 319275.01.
Average BasePay: 67207.55842464653: This shows the average or mean value of all the base pay values in the DataFrame. The average base pay is approximately 67207.56.


6. Replace 'Not provided' in EmployeeName Column to Nan


In [ ]:
df['EmployeeName'] = df['EmployeeName'].replace('Not provided', np.nan)

<ipython-input-46-65d220a2215d>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['EmployeeName'] = df['EmployeeName'].replace('Not provided', np.nan)


In [ ]:
df

EmployeeName                      JobTitle    BasePay  \
36159       Gary Altenberg  Lieutenant, Fire Suppression  128808.87   
36160         Gregory Suhr               Chief of Police  302578.00   
36161           Khoa Trinh   Electronic Maintenance Tech  111921.00   
36162   Joanne Hayes-White        Chief, Fire Department  296943.01   
36163    Frederick Binkley     EMT/Paramedic/Firefighter  126863.19   
...                    ...                           ...        ...   
148645    Carolyn A Wilson     Human Services Technician       0.00   
148647      Joann Anderson   Communications Dispatcher 2       0.00   
148648         Leon Walker                     Custodian       0.00   
148649       Roy I Tillery                     Custodian       0.00   
148653           Joe Lopez    Counselor, Log Cabin Ranch       0.00   

        OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  
36159     220909.48  13126.31  44430.12  362844.66         407274.78  2012  
36160          0.00  18974.11  69810.19  321552.11         391362.30  2012  
36161     146415.32  78057.41  53102.29  336393.73         389496.02  2012  
36162          0.00  17816.59  72047.88  314759.60         386807.48  2012  
36163     192424.49  17917.18  44438.25  337204.86         381643.11  2012  
...             ...       ...       ...        ...               ...   ...  
148645         0.00      0.00      0.00       0.00              0.00  2014  
148647         0.00      0.00      0.00       0.00              0.00  2014  
148648         0.00      0.00      0.00       0.00              0.00  2014  
148649         0.00      0.00      0.00       0.00              0.00  2014  
148653         0.00   -618.13      0.00    -618.13           -618.13  2014  

[111886 rows x 9 columns]

7. Drop The Rows Having 5 Missing Values


In [ ]:
df = df.dropna(thresh=len(df.columns) - 4)

In [ ]:
df

EmployeeName                      JobTitle    BasePay  \
36159       Gary Altenberg  Lieutenant, Fire Suppression  128808.87   
36160         Gregory Suhr               Chief of Police  302578.00   
36161           Khoa Trinh   Electronic Maintenance Tech  111921.00   
36162   Joanne Hayes-White        Chief, Fire Department  296943.01   
36163    Frederick Binkley     EMT/Paramedic/Firefighter  126863.19   
...                    ...                           ...        ...   
148645    Carolyn A Wilson     Human Services Technician       0.00   
148647      Joann Anderson   Communications Dispatcher 2       0.00   
148648         Leon Walker                     Custodian       0.00   
148649       Roy I Tillery                     Custodian       0.00   
148653           Joe Lopez    Counselor, Log Cabin Ranch       0.00   

        OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  
36159     220909.48  13126.31  44430.12  362844.66         407274.78  2012  
36160          0.00  18974.11  69810.19  321552.11         391362.30  2012  
36161     146415.32  78057.41  53102.29  336393.73         389496.02  2012  
36162          0.00  17816.59  72047.88  314759.60         386807.48  2012  
36163     192424.49  17917.18  44438.25  337204.86         381643.11  2012  
...             ...       ...       ...        ...               ...   ...  
148645         0.00      0.00      0.00       0.00              0.00  2014  
148647         0.00      0.00      0.00       0.00              0.00  2014  
148648         0.00      0.00      0.00       0.00              0.00  2014  
148649         0.00      0.00      0.00       0.00              0.00  2014  
148653         0.00   -618.13      0.00    -618.13           -618.13  2014  

[111886 rows x 9 columns]

8. Find The Job Title Of ALBERT PARDINI


In [ ]:
job_title = df[df['EmployeeName'] == 'ALBERT PARDINI']['JobTitle'].values[0]
print(f"The job title of ALBERT PARDINI is: {job_title}")

IndexError: index 0 is out of bounds for axis 0 with size 0

In [ ]:
if 'ALBERT PARDINI' in df['EmployeeName'].values:
    print("ALBERT PARDINI is present in the data.")
else:
    print("ALBERT PARDINI is not present in the data.")

9. How Much ALBERT PARDINI makes(include Benefits)


10. Display Name Of The Person Having The Highest Basepay


In [ ]:
highest_basepay_person = df.loc[df['BasePay'].idxmax(), 'EmployeeName']
print(f"The person with the highest BasePay is: {highest_basepay_person}")

11. Find The average Basepay of all Employees per Year


In [ ]:
average_basepay_per_year = df.groupby('Year')['BasePay'].mean()
print(average_basepay_per_year)

12. Find Average Basepay Of All Employee Per Job Title


In [ ]:
average_basepay_per_job = df.groupby('JobTitle')['BasePay'].mean()
print(average_basepay_per_job)

13. Find The Average Basepay of All Employee Having job Title Accountant

In [ ]:
average_basepay_accountant = df[df['JobTitle'] == 'Accountant']['BasePay'].mean()
print(f"Average BasePay for Accountants: {average_basepay_accountant}")

14. Five Most Common Job

In [ ]:
most_common_jobs = df['JobTitle'].value_counts().head(5)
print(most_common_jobs)